File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
from collections import deque

def solution(bridge_length, weight, truck_weights):
    answer = 0
    trucks = deque([0]*bridge_length, maxlen = bridge_length)
    truck_weights = deque(truck_weights)
    bridge_weight = 0
    
    while truck_weights: # 트럭 다 올리기
        nxt = truck_weights[0]
        out = trucks[-1]
        if nxt > weight:
            truck_weights.popleft()
            continue
        answer += 1
        if bridge_weight - out + nxt <= weight: # 올라갈 수 있음
            truck_weights.popleft() # 대기 트럭에서 빼고
            trucks.appendleft(nxt) # 다리 위에 올림
            bridge_weight = bridge_weight - out + nxt
        else:
            trucks.appendleft(0)
            bridge_weight -= out

    return answer + bridge_length

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.*;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        int answer = 0;
        Deque<Integer> bridge = new ArrayDeque<>();
        int total_weight = 0;
        int index = 0;
        
        while (index < truck_weights.length) {
            if (bridge.size() == bridge_length)
                total_weight -= bridge.poll();
            
            // 현재 트럭이 다리에 오를 수 있다면
            if (total_weight + truck_weights[index] <= weight) {
                bridge.offer(truck_weights[index]);
                total_weight += truck_weights[index];
                index++;
            }
            else // 오를 수 없다면 빈 공간으로 채움
                bridge.offer(0);
           
            answer++;
        }

        return answer + bridge_length;
    }
}